core/arch/arm/plat-k3/drivers/ti_sci.c

```diff
@@ -134,24 +144,32 @@ static inline int ti_sci_get_response(struct ti_sci_xfer *xfer)
 static inline int ti_sci_do_xfer(struct ti_sci_xfer *xfer)
 {
 	struct k3_sec_proxy_msg *msg = &xfer->tx_message;
+	struct ti_sci_msg_hdr *hdr = NULL;
 	int ret = 0;
 
+	mutex_lock(&ti_sci_mutex_lock);
+
+	hdr = (struct ti_sci_msg_hdr *)msg->buf;
+	current_sequence = hdr->seq;
```

Comment:
We could set the `hdr->seq` here instead of up in `ti_sci_setup_xfer()`, that would put it inside this mutex section and remove the need for the extra spinlock. Then as both updating and checking against message_sequence happen in this same mutex section, you'll also not need the extra temp `current_sequence`, just use `message_sequence` in `ti_sci_get_response()` as before.

Reply:
Hey @glneo , that seems like a good idea but in general wondering something.. I see that `message_sequence` is a uint8_t datatype which means that the maximum limit of messages should be around 255 before it rolls over and comes back to 0. 

Am guessing that we do more message communications than this and we must've been rolling over and coming to 0? If yes, do we even need a counter if we are keeping this code under mutex? ( Every call will be sequential so we shouldn't require it am guessing and we can keep the hdr->seq = 0 always? )

Reply:
Right, `message_sequence` rolls over every now and then, it is just a check for us that the message we get back is probably in response to the message we sent out if the number matches. The actual value is unimportant, just that it is different from the messages around it. It could be a random number generated each time too.

In theory, each send and receive call pair should be inside a mutex together so this shouldn't matter. In practice I have been saved with this check in some odd cases where a message was stuck in the SProxy queue and even though I just sent a message, I was getting a response for a message sent before OP-TEE even started.

Reply:
Oh okay, then am wondering that the check should be at host_id level rather than message_sequence right?

Reply:
Are you asking if we only need to check `host_id`? That would prevent messages from other hosts in our queue like my above example, yes, but still not prevent any issues in OP-TEE from being detected. I'd like to point out the issue that lead to this PR was discovered because we had this check in place. Could you imaging how hard it would have been to debug if we didn't have the check and we simply processed the returned message as if it was the correct order..

If we are confident the mutex is enough to prevent any future bugs then I'd be okay with dropping the sequence check.

Reply:
Yeah don't mind keeping the check for sanity purpose, just wondered if can add host_id check as well then as that might make future issues more clear incase they arise, would update the PR. Thanks!

Reply:
@glneo  updated